### PR TITLE
Add tests for cleaning policies operation

### PR DESCRIPTION
**Pull request by rafalste**

### CODE REVIEW

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+def test_cleaning_policies_in_write_through(cleaning_policy):
+    """
+        title: Test for cleaning policy operation in Write-Through cache mode.
+        description: |
+          Check if ALRU, NOP and ACP cleaning policies preserve their
+          parameters when changed and if they flush dirty data properly
+          in Write-Through cache mode.
+        pass_criteria:
+          - Flush parameters are set to different values if applied.
+          - Dirty data are flushed or not according to policy used.
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+def test_cleaning_policies_in_write_back(cleaning_policy):
+    """
+        title: Test for cleaning policy operation in Write-Back cache mode.
+        description: |
+          Check if ALRU, NOP and ACP cleaning policies preserve their
+          parameters when changed and if they flush dirty data properly
+          in Write-Back cache mode.
+        pass_criteria:
+          - Flush parameters are set to different values if applied.
+          - Dirty data are flushed or not according to policy used.
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_cleaning_policies_in_write_back(cleaning_policy):
+    """
+        title: Test for cleaning policy operation in Write-Back cache mode.
+        description: |
+          Check if ALRU, NOP and ACP cleaning policies preserve their
+          parameters when changed and if they flush dirty data properly
+          in Write-Back cache mode.
+        pass_criteria:
+          - Flush parameters are set to different values if applied.
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Not very clear criterium - I had to guess what you had in mind

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+          - Dirty data is flushed or not according to policy used.
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: the policy

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        Udev.enable()
+
+
+def storage_prepare():
+    cache_dev = TestRun.disks["cache"]
+    cache_dev.create_partitions([Size(1, Unit.GibiByte)])
+    core_dev = TestRun.disks["core"]
+    parts = []
+    for i in range(cores_count):
+        parts.append(Size(2, Unit.GibiByte))
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: parts = [Size(2, Unit.GibiByte)] *  cores_count

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+                "core writes should equal total IO on cache."
+            )
+    if cleaning_policy == CleaningPolicy.nop:
+        if core_writes_after_cleaning.value != 0 or core_writes_before_cleaning.value != 0:
+            TestRun.fail(
+                "NOP cleaning policy is not working properly! "
+                "There should be no core writes as there is no cleaning of dirty data."
+            )
+    if cleaning_policy == CleaningPolicy.acp:
+        if core_writes_after_cleaning <= core_writes_before_cleaning:
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: I'd add a check here that core_writes_before_cleaning != 0
BTW if acp is cleaning all the time and nop is never cleaning, then before/after_cleaning are not exactly right names for variables ;)

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py (hunk 3 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py (hunk 4 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py (hunk 5 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py (hunk 6 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        for i in range(cores_count):
+            core.append(cache.add_core(core_dev.partitions[i]))
+
+    with TestRun.step("Change cache mode to Write-Back"):
+        cache.set_cache_mode(CacheMode.WB)
+
+    with TestRun.step("Run 'fio'"):
+        for i in range(cores_count):
+            fio_prepare(core[i]).run()
+        sync()
```
Comment (robertbaldyga) [inline test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: This `sync()` could potentially take a long time depending on DUT state, so it's technically possible that cleaning may start before we check initial core stats, effectively causing test to fail. The simplest solution would be to check stats before calling `sync()`, just after running `fio`.

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+        core = []
+        for i in range(cores_count):
+            core.append(cache.add_core(core_dev.partitions[i]))
+
+    with TestRun.step("Change cache mode to Write-Back"):
+        cache.set_cache_mode(CacheMode.WB)
+
+    with TestRun.step("Run 'fio'"):
+        for i in range(cores_count):
+            fio_prepare(core[i]).run()
```
Comment (robertbaldyga) [inline test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: We may want to run `fio` in parallel on all cores.

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py (hunk 7 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: So is this sync() needed then? Maybe I will just get rid of it?

Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py (hunk 8 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

File: test/functional/tests/cache_ops/test_cleaning_policy_operation.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+                failed_params += (
+                    f"Flush Max Buffers is {current_alru_params.flush_max_buffers}, "
+                    f"should be {alru_params.flush_max_buffers}\n"
+                )
+            if current_alru_params.activity_threshold != alru_params.activity_threshold:
+                failed_params += (
+                    f"Activity Threshold is {current_alru_params.activity_threshold}, "
+                    f"should be {alru_params.activity_threshold}\n"
+                )
+            TestRun.fail(f"ALRU parameters did not switch properly:\n{failed_params}")
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: I'm not sure if we should stop the test when it fails (and TestRun.fail does that) - consider using TestRun.LOGGER.error (in the whole file)

Comment (katlapinka) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: I believe, that in test step called "Check for running CAS cleaner" there should be TestRun.failed as it is, but like Daniel said, in other steps you can change it to the TestRun.LOGGER.error :)

Comment (Deixx) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Maybe Kasia is right - I thought about situation, when for some reason cleaner process name changes (then this check would abort test), but after giving it some thought I think that unexpected name could be a test-stopper ;)

Comment (rafalste) [reply test/functional/tests/cache_ops/test_cleaning_policy_operation.py]: Done

**Pull request merged**